docs/madr/decisions/006-mesh-traffic-permission.md

```diff
@@ -0,0 +1,576 @@
+
+Technical Story: https://github.com/kumahq/kuma/issues/4222
+
+## Context and Problem Statement
+
+[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
+Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
+Current MADR aims to define a MeshTrafficPermission.
+
+## Considered Options
+
+* Create a MeshTrafficPermission policy
+
+## Decision Outcome
+
+Chosen option: create a MeshTrafficPermission policy
+
+### Overview
+
+MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
+There are might be several reasons why it could be useful to block unwanted traffic:
+* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)
+
+### Specification
+
+Top-level targetRef can have all available kinds:
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshTrafficPermission is an inbound policy that's why it has a "from" array:
+
+```yaml
+from:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+      name: ...
+```
+
+Configuration for MeshTrafficPermission has 2 fields "action" and "shadow_action" with values "ALLOW" or "DENY".
+So the overall specification looks like this:
+
+```yaml
+type: MeshTrafficPermission
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+    name: ...
+  from:
+    - targetRef:
+        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+        name: ...
+      default:
+        action: ALLOW|DENY
+        shadow_action: ALLOW|DENY
+```
+
+We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
+The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 
+
+### Shadow actions
+
+Shadow actions are not enforced but emit stats and can be used for rule testing.
+For example, before creating "action: DENY" rule, user creates "shadow_action: DENY" rule and checks the stats.
+If there are no unexpected denied requests from the clients then it's safe to change "shadow_action" to "action".
+
+Alternatively we could introduce `shadow: true|false` instead of `shadow_action: ALLOW|DENY`. 
+But this approach has its flaws:
+* policy matching makes it less obvious with merging. It's easy to create a rule like:
+    ```yaml
+    targetRef:
+       kind: Mesh
+    default:
+      action: DENY
+      shadow: true
+   ```
+   and unintentionally shadow all less specific rules.
+
+* Envoy configuration has 2 independent places for "rules" and "shadow_rules", 
+so having independent fields "action" and "shadow_action" makes it easier to translate to Envoy configs.
```

Comment:
This doesn't make sense to me. is it hard either way? Why is one way easier?

Reply:
If you have `action: ALLOW|DENY` and `shadow: true|false`, then these fields are sort of correlated (changing of the `shadow` flag changes the meaning of `action`). You can't independently generate "rules" and "shadow_rules", at least it's not simple.

But if you have `action: ALLOW|DENY` and `shadow_action: ALLOW|DENY`, then these fields are not correlated. And you can generate "rules" and "shadow_rules" independently. And that's exactly how Envoy has it in the configuration.

Does it make sense? I can rephrase the sentence in the document

Reply:
You can always convert the `{shadow: bool, action: ALLOW|DENY}` model into the `{shadowAction: ALLOW|DENY, action: ALLOW|DENY}` model, so it isn't any simpler or harder, right?

Do you mean that you can't express something like `{action: ALLOW, shadowAction: DENY}` in one entry with the `shadow: bool` model?

Reply:
In fact I agree with @jakubdyszkiewicz that `ALLOW|DENY|SHADOW_ALLOW|SHADOW_DENY` would be the simplest. `{action, shadowAction}` is too unconstrained in my opinion and allows easily misconfiguring something like `{action: DENY, shadowAction: DENY}`

Reply:
That's true, you can convert, but the boolean flag makes it difficult when merging. For example:
```yaml
type: MeshTrafficPermission
name: global
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
         kind: MeshService
         name: backend
       default:
         action: DENY
---
type: MeshTrafficPermission
name: mtp-web
spec:
  targetRef:
    kind: MeshService
    name: web
  from:         
    - targetRef:
         kind: Mesh
       default:
         action: DENY
         shadow: true
```
in this example `shadow: true` from `mtp-web` changes the behaviour of `action: DENY` in `global`. And in fact you have no actual DENY rules in this case, only shadow rules

Reply:
But how does that make it "difficult", it's just easy to misconfigure, right? That's the same as the following "unintentionally shadow rules" point.

Reply:
that's just `action: SHADOW_DENY`, isn't it?

Reply:
Resolving this in favor of my third comment where we end up discussing the same topic.